fvm/environment/minimum_required_version.go

```diff
@@ -0,0 +1,67 @@
+package environment
+
+import (
+	"github.com/coreos/go-semver/semver"
+
+	"github.com/onflow/flow-go/fvm/storage/state"
+)
+
+// MinimumRequiredVersion returns the minimum required cadence version for the current environment
+// in semver format.
+type MinimumRequiredVersion interface {
+	MinimumRequiredVersion() (string, error)
+}
+
+type minimumRequiredVersion struct {
+	txnPreparer state.NestedTransactionPreparer
+}
+
+func NewMinimumRequiredVersion(
+	txnPreparer state.NestedTransactionPreparer,
+) MinimumRequiredVersion {
+	return minimumRequiredVersion{
+		txnPreparer: txnPreparer,
+	}
+}
+
+func (c minimumRequiredVersion) MinimumRequiredVersion() (string, error) {
```

Comment:
```suggestion
// The returned cadence version can be used by cadnece runtime for supporting feature flag. 
// The feature flag in cadence allows ENs to produce consistent results even if running with 
// different cadence versions at the same height, which is useful for rolling out cadence 
// upgrade without all ENs restarting all together.
// For instance, we would like to grade cadence from v1 to v3, where v3 has a new cadence feature.
// We first make a cadence v2 that has feature flag only turned on when the MinimumRequiredVersion()
// method returns v2 or above. 
// So cadence v2 with the feature flag turned off will produce the same result as v1 which doesn't have the feature.
// And cadence v2 with the feature flag turned on will also produce the same result as v3 which has the feature.
// The feature flag allows us to roll out cadence v2 to all ENs which was running v1. 
// And we use the MinimumRequiredVersion to control when the feature flag should be switched from off to on.
// And the switching should happen at the same height for all ENs.
// 
// The height-based switch over can be done by using VersionBeacon, however, the VersionBeacon only 
// defines the flow-go version, not cadence version.
// So we first read the current minimum required flow-go version from the VersionBeacon control, 
// and map it to the cadence version to be used by cadence to decide feature flag status.
//
// For instance, let’s say all ENs are running flow-go v0.37.0 with cadence v1.
// We first create a version mapping entry for flow-go v0.37.1 to cadence v2, and roll out v0.37.1 to all ENs.
// v0.37.1 ENs will produce the same result as v0.37.0 ENs, because the current version beacon still returns v0.37.0,
// which maps zero cadence version, and cadence will keep the feature flag off.
// 
// After all ENs have upgraded to v0.37.2, we send out a version beacon to switch to v0.37.1 at a future height,
// let’s say height 1000.
// Then what happens is that:
// 1. ENs running v0.37.0 will crash after height 999, until upgrade to higher version
// 2. ENs running v0.37.1 will execute with cadence v2 with feature flag off up until height 999, and from height 1000,
//     the feature flag will be on, which means all v0.37.1 ENs will again produce consistent results for blocks above 1000.
// After height 1000 have been sealed, we can roll out v0.37.2 to all ENs with cadence v3, and it will produce the consistent
// result as v0.37.1.
func (c minimumRequiredVersion) MinimumRequiredVersion() (string, error) {
```